File: src/api/providers.ts
import { apiClient } from "./client";

export interface Provider {
  id: string;
  name: string;
  protocol: "ollama" | "openai";
  base_url: string;
  requires_key: boolean;
  has_key: boolean;
  masked_key: string;
  enabled: boolean;
  local: boolean;
  default_models: string[];
}

export interface UpsertProviderRequest {
  id: string;
  name?: string;
  base_url?: string;
  /** Plaintext key — sent to the backend (encrypted there), never persisted locally. */
  api_key?: string;
  enabled: boolean;
}

export const providerApi = {
  async list(): Promise<{ providers: Provider[]; encryption_available: boolean }> {
    const { data } = await apiClient.get("/providers");
    return data;
  },

  async upsert(request: UpsertProviderRequest): Promise<Provider> {
    const { data } = await apiClient.post("/providers", request);
    return data;
  },

  async test(providerId: string): Promise<{ ok: boolean; status?: number; error?: string }> {
    const { data } = await apiClient.post(`/providers/${providerId}/test`);
    return data;
  },

  async remove(providerId: string): Promise<{ success: boolean }> {
    const { data } = await apiClient.delete(`/providers/${providerId}`);
    return data;
  },
};

File: backend/routers/providers.py
"""Provider management: list, add/update, test, delete LLM providers."""

import logging
from typing import Optional

from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field

from dependencies import get_provider_service
from errors import NotFoundError
from services.provider_service import ProviderService

logger = logging.getLogger(__name__)
router = APIRouter()


class UpsertProviderRequest(BaseModel):
    # The preset id (openai/openrouter/groq/…/custom). For "custom", a base_url
    # is required and a unique id may be supplied.
    id: str = Field(..., min_length=1)
    name: Optional[str] = None
    base_url: Optional[str] = None
    api_key: Optional[str] = None
    enabled: bool = True


@router.get("")
async def list_providers(
    service: ProviderService = Depends(get_provider_service),
):
    """All providers (presets + stored), with masked keys. Safe for the UI."""
    return {
        "providers": await service.list_providers(),
        "encryption_available": service.encryption_available,
    }


@router.post("")
async def upsert_provider(
    request: UpsertProviderRequest,
    service: ProviderService = Depends(get_provider_service),
):
    """Add or update a provider. The API key (if given) is encrypted at rest."""
    return await service.upsert_provider(
        request.id,
        name=request.name,
        base_url=request.base_url,
        api_key=request.api_key,
        enabled=request.enabled,
    )


@router.post("/{provider_id}/test")
async def test_provider(
    provider_id: str,
    service: ProviderService = Depends(get_provider_service),
):
    """Probe the provider to confirm it's reachable and the key works."""
    return await service.test_provider(provider_id)


@router.delete("/{provider_id}")
async def delete_provider(
    provider_id: str,
    service: ProviderService = Depends(get_provider_service),
):
    """Remove a stored provider override (presets remain available)."""
    if not await service.delete_provider(provider_id):
        raise NotFoundError("Provider not found")
    return {"success": True}
